src/movement.rs

```diff
@@ -50,45 +66,102 @@ pub fn player_controller(
+                dir = dir * stats.movement_speed * time.delta_seconds();
 
-        //Restrict player to the ground
-        let new_y = transform.translation.y + dir.y + consts::GROUND_OFFSET;
+                let new_x = transform.translation.x + dir.x;
 
-        if new_y >= consts::MAX_Y || new_y <= consts::MIN_Y {
-            dir.y = 0.;
-        }
+                // The dir.x condition allows some flexibility (e.g. in case of knockback), given
+                // the current state of development. To be removed once the movement logic is
+                // stabilized.
+                //
+                if dir.x < 0. && new_x < left_movement_boundary.0 {
+                    dir.x = 0.;
+                }
+                println!("{}", new_x);
+                //Restrict player to the ground
+                let new_y = transform.translation.y + dir.y + consts::GROUND_OFFSET;
+
+                if new_y >= consts::MAX_Y || new_y <= consts::MIN_Y {
+                    dir.y = 0.;
+                }
+
+                //Move the player
+                Some(dir)
+            }
+        })
+        .collect::<Vec<_>>();
+
+    if player_dirs.len() == 2 {
+        let window_width = windows.get_primary().unwrap().width();
+        let max_players_x_distance = window_width * game_meta.max_players_x_pct_distance;
 
-        //Move the player
-        transform.translation.x += dir.x;
-        transform.translation.y += dir.y;
+        let new_players_x = players_x
+            .iter()
+            .zip(player_dirs.iter())
+            .map(|(x, dir)| x + dir.unwrap_or(Vec2::ZERO).x)
+            .collect::<Vec<_>>();
 
-        //Set the player state and direction
-        if let Some(mut facing) = facing_option {
-            if dir.x < 0. {
-                facing.set(Facing::Left);
-            } else if dir.x > 0. {
-                facing.set(Facing::Right);
+        if (new_players_x[0] - new_players_x[1]).abs() > max_players_x_distance {
+            let right_player_i = if new_players_x[0] > new_players_x[1] {
+                0
+            } else {
+                1
+            };
+
+            // We could clamp the dir.x, however, if we do that, we'd have to deal with jitter.
+            // Since we assume that the movement in each frame is small, we just stop the movement.
+            //
+            if player_dirs[right_player_i].unwrap().x > 0. {
+                player_dirs[right_player_i] = Some(Vec2::ZERO);
             }
         }
+    }
 
-        if dir == Vec2::ZERO {
-            state.set(State::Idle);
-        } else {
-            state.set(State::Running);
+    for ((mut state, _, mut transform, facing_option, _), dir) in
+        query.iter_mut().zip(player_dirs.iter())
+    {
+        if let Some(dir) = dir {
+            transform.translation.x += dir.x;
+            transform.translation.y += dir.y;
+
+            //Set the player state and direction
+            if let Some(mut facing) = facing_option {
```

Comment:
I must have missed it in a different merge, I don't think facing needs to be an option at all in this system. Could be changed in this pr or left to another. @zicklag I could be wrong if there is a way the default playerbundle could ever not be added.

Reply:
Yeah, as far as I can tell I don't see any reason the facing component wouldn't be there.

Reply:
removed in https://github.com/fishfight/punchy/commit/9629e92a3eba32bb6cbfb5fe11d22c87429348da